src/Passes/libs/OpsCounter/OpsCounter.hpp

```diff
@@ -0,0 +1,69 @@
+#pragma once
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT License.
+
+#include "Llvm.hpp"
+
+namespace microsoft
+{
+namespace quantum
+{
+
+    class OpsCounterAnalytics : public llvm::AnalysisInfoMixin<OpsCounterAnalytics>
+    {
+      public:
+        using Result = llvm::StringMap<unsigned>;
+
+        /// Constructors and destructors
+        /// @{
+        OpsCounterAnalytics()                           = default;
+        OpsCounterAnalytics(OpsCounterAnalytics const&) = delete;
+        OpsCounterAnalytics(OpsCounterAnalytics&&)      = default;
+        ~OpsCounterAnalytics()                          = default;
+        /// @}
+
+        /// Operators
+        /// @{
+        OpsCounterAnalytics& operator=(OpsCounterAnalytics const&) = delete;
+        OpsCounterAnalytics& operator=(OpsCounterAnalytics&&) = delete;
+        /// @}
```

Comment:
(I have no experience with this (`llvm::`) code base, feel free to ignore if my note is not applicable)

It looks somewhat suspicious that for copy functions (copy constructor, copy assignment operator) both are `delete`d,
but for the move functions (move constructor and move assignment operator) one is `default` and the other is `delete`.

I also have an impression that explicitly providing the destructor (`~OpsCounterAnalytics() = default;`) but deleting some other special member functions somewhat violates "the rule of all or nothing".

Just in case some more info:
* [Item 5: Know what functions C++ silently writes and calls](http://wavelino.coffeecup.com/pdf/Effective%20Cshapshap.pdf)
* What is move semantics? See _both_ [here](https://stackoverflow.com/a/3109981/6362941) and [here](https://stackoverflow.com/a/11540204/6362941)
* [Item 17: Understand special member function generation](https://moodle.ufsc.br/pluginfile.php/2377667/mod_resource/content/0/Effective_Modern_C__.pdf)
* [Know What Happens If Rule of Three Is Violated](https://github.com/kuzminrobin/code_review_notes/blob/master/cpp_design_bookmarks.md#know-what-happens-if-rule-of-three-is-violated)

Reply:
I am not entirely sure that I understand this comment, so if the answer is a bit off, please ask again.

As such there is nothing wrong in deleting the copy constructor and copy assignment while preserving the move constructor and move assignment. This is the typical behaviour of a file or network connection where you do not want to make a copy of the actual connection, but you do want the ability to do the construction in one part of the code and use it in another part.

The specific configuration chosen here is LLVM specific and dictated by the parent class - the code just explicitly declares this behaviour such that it is apparent to any one who uses the code that this is the behaviour of the class. With any other configuration (using delete/default) your compiler will complain that it is not compatible with the parent class. The only thing that imo is a bit odd about the configuration is having a move constructor without a move assignment. This may be due to some optimisation or to enforce a specific behaviour in the LLVM library.

When you refer to the "rule of all or nothing", you mean the rule of three/five? Note that rule-of-X only applies to user-defined constructors/destructors/operations:

> If a class requires a user-defined destructor, a user-defined copy constructor, or a user-defined copy assignment operator, it almost certainly requires all three.

See [rule of three](https://en.cppreference.com/w/cpp/language/rule_of_three)

There is no user-defined constructors/destructors/operations in the the code. The code merely states that we are using the default or prohibiting the use of default (the ones the compiler would have used anyway) constructors/destructor/operations. My personal preference is that all constructors are explicitly declared every time. This makes it easy to see what the class can and cannot do, hence, the explicit declarations. In this way, it makes it easier to read the code, as you immediately know that this class will only default construct and move construct (without having to read the LLVM code base). As such you could just delete all of these and the class would fall back to using its parents constructors/operators/destructor.

Please let me know if this did not answer your concern.

Reply:
This fragment 
> The only thing that imo is a bit odd about the configuration is _having a move constructor without a move assignment_. This may be due to some optimisation or to enforce a specific behaviour in the LLVM library.

answers the first half of my note 
(In my note I shouldn't have mentioned the copy constructor and copy assignment, I didn't mean to discuss them, I'm all right that they are deleted. I just mentioned them for comparison with the move constructor and move assignment, one of which is deleted and the other one is not, which looked strange).

As for the second half of my note - the destructor... The topic is rather tricky and I'm not sure I should have mentioned that. Any way I consider this discussion as resolved.

If you are interested then I can try to explain what I was thinking about in the second half.
(All the special member functions are mentioned explicitly, that is why the "rule of all or nothing" (aka "rule of 3", "rule of 5", "rule of 0") should be ok here, I shouldn't have mentioned that rule, sorry, my bad)

As for the fragment `~OpsCounterAnalytics() = default;`. I believe it provides the user-declared destructor and makes a difference from a situation where this fragment is skipped. I played with the examples a bit but failed to create an example that demonstrates the difference. Looks like the required code has been optimized out in my examples.

Sorry for taking your time. Marking as resolved.